### ISSUE TITLE
Typed error handling

### ISSUE THREAD

**Issue by fgnass** [cellular/oazapfts#20]
Dear oazapfts users,

I'd like to hear your opinion on the way we should handle errors in future. Currently the first response in the spec with a code `< 400` is used as return type. All responses with a code `>= 400` throw a `HttpError` with `{ status: number; data: any }`

## Option 1: All responses are resolved

Instead of throwing, all responses are resolved and each method returns a union type of all possible responses. API consumers must add a **type-guard** on the status in order to access the correctly typed response:

```ts
const res = await login("me@example.com", "secret");
switch (res.status) {
  case 200:
    const { id, accessToken } = res.data;
    // Handle success case here
    break;
  case 400:
    const { fieldErrors } = res.data;
    // Handle the typed error here
    break;
}
```

## Option 2: Error handler

Each function gets a `handleError` property that can be used in the catch-block and resolves the correctly typed error payload:

```ts
try {
  const { id, accessToken } = await login("me@example.com", "secret");
  // Handle success case here
} catch (err) {
  login.handleError(err, {
    400: ({fieldErrors}) => {
      // Handle the typed error here
    },
  });
}
```

## Please vote

Option 1 would be a breaking change as the success response can no longer be destructured right away. Also note that you would still need a try/catch block to handle network errors.
Option 2 is sugar on top of the current model but maybe a little uncommon.

Please use the following emojis to vote and feel welcome to leave a comment:

👀 Option 1 - All responses are resolved
🚀 Option 2 - Error handler

**Comment by Xiphe** [comment]
I don't have a strong opinion here. I Like both approaches for different reasons. 
1 feels more natural to me
2 is probably more useful once you internalised it

If we go with 2 i'd like to add following criteria.

- `handleError` should pass-through the return value of error handlers
- `handleError` must not catch errors thrown in error handlers
- it should be required to add a default error handler to handle other errors 
   - argument is the bare error passed into it.
   - default handler is called when the error is unknown to the api or no handler is defined for that error

```ts
function login(name: string, pass: string): string {
  try {
    if (Math.random() > 0.5) { throw new Error('other'); }
    const { id, accessToken } = await login("me@example.com", "secret");
    return 'OK';
  } catch (err) {
    return login.handleError(err, {
      400: ({fieldErrors}) => (fieldErrors),
      500: ({serverError}) => throw new Error(serverError),
      default: (e2) => {
        console.assert(e2 === err);
        return "Something went wrong";
      }
    });
  }
}
```

**Comment by fgnass** [comment]
One potential drawback of Option 2 is that if a function defines **multiple** response codes below `400` we can only return a union type of all of them. Since the status code is not exposed in the success case we can't use it as discriminator.

For option 1 it would also be possible to create a generic `handle` utility that infers the types correctly:

```ts
await handle(login("me@example.com", "secret"), {
   200: ({id, accessToken}) => "OK",
   400: ({fieldErrors}) => fieldErrors
}))
```

Or another one that brings back the current behaviour of throwing on errors:

```ts
const {id, accessToken} = await ok(login("me@example.com", "secret"));
```

**Comment by Xiphe** [comment]
but the `handle` would have the same drawback as option 1, wouldn't it?

**Comment by mostheimer** [comment]
I would prefer option 1, along with generic utilities to handle the errors. 
But I would suggest to handle network errors as well, so no double logics have to be implemented. 
Either in a default case, with untyped errors, or in an extra case. (status="networkError"?)

**Comment by fgnass** [comment]
I created a [next branch](https://github.com/cellular/oazapfts/tree/next) that implements option 1 with optional `ok` and `handle` helpers. It is still work-in-progress but feels like the right approach to me. The `oazapfts` package is now a runtime dependency as I moved all the runtime code from the generated file to `oazapfts/runtime`. The two helper functions can be imported from the top-level `oazapfts` module.
